bluemira/base/reactor_config.py

```diff
@@ -0,0 +1,234 @@
+
+from bluemira.base.error import ReactorConfigError
+from bluemira.base.look_and_feel import bluemira_warn
+from bluemira.base.parameter_frame import ParameterFrame, make_parameter_frame
+
+
+@dataclass
+class ConfigParams:
+    """
+    Class to hold the params return by `params_for`
+    from a `ReactorConfig` object
+    """
+
+    global_params: ParameterFrame
+    local_params: Dict
+
+
+_PfT = TypeVar("_PfT", bound=ParameterFrame)
+_PARAMETERS_KEY = "params"
+
+
+class ReactorConfig:
+    """
+    Class that provides a simple interface over config JSON files and
+    handles overwriting multiply defined attributes.
+
+    If an attribute is defined more than once in a component,
+    the more globally scoped value is used (global overwrites local).
+
+    Parameters
+    ----------
+    config_path: str | dict
+        The path to the config JSON file or a dict of the data.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        @dataclass
+        class GlobalParams(ParameterFrame):
+            a: Parameter[int]
+
+
+        reactor_config = ReactorConfig(
+            {
+                "params": {"a": 10},
+                "comp A": {
+                    "params": {"a": 5, "b": 5},
+                    "designer": {
+                        "params": {"a": 1},
+                        "some_config": "some_value",
+                    },
+                    "builder": {
+                        "params": {"b": 1, "c": 1},
+                        "another_config": "another_value",
+                    },
+                },
+                "comp B": {
+                    "params": {"b": 5},
+                    "builder": {
+                        "third_config": "third_value",
+                    },
+                },
+            },
+            GlobalParams,
+        )
+
+    """
+
+    @staticmethod
+    def _read_json_file(path: str) -> dict:
+        with open(path, "r") as f:
+            return json.load(f)
+
+    def __init__(
+        self,
+        config_path: Union[str, dict],
+        global_params_type: Type[_PfT],
+    ):
+        if isinstance(config_path, str):
+            self.config_data = ReactorConfig._read_json_file(config_path)
+        elif isinstance(config_path, dict):
+            self.config_data = config_path
+        else:
+            raise ReactorConfigError("Invalid config_path")
+
+        self.global_params: _PfT = make_parameter_frame(
+            self.config_data.get(_PARAMETERS_KEY, {}),
+            global_params_type,
+        )
+
+    def __str__(self) -> str:
+        """Returns config_data as a nicely pretty formatted string"""
+        return self._pprint_dict(self.config_data)
+
+    def _pprint_dict(self, d: dict) -> str:
+        return pprint.pformat(d, sort_dicts=False, indent=1)
+
+    def _warn_on_duplicate_keys(self, shared_key: str, arg: str):
+        bluemira_warn(f"'{shared_key}' is already defined for {arg}")
```

Comment:
Could we put a little more detail in this warning? This could be a source of subtle bugs in user code, so we should be as specific as possible to help them out. The warning here says a key is already defined, but not what action we took to mitigate this - they may assume the lower scope overrides the higher.

I think we should at least say which key we are ignoring

Reply:
100%